src/Socket/Stream/IPv4.hs

```diff
@@ -1280,6 +1283,47 @@ receiveBetweenMutableByteArraySlice !conn@(Connection fd) !minBytes !maxBytes !m
       functionReceiveBetweenMutableByteArraySlice
       [SCK.negativeSliceLength]
 
+interruptibleReceiveBoundedByteStringSlice ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> Int -- ^ Maximum number of bytes to receive
+  -> Int -- ^ Offset into the buffer
+  -> IO (Either (ReceiveException 'Interruptible) ByteString) -- ^ Either a socket exception or the number of ByteString we allocated
+interruptibleReceiveBoundedByteStringSlice !abandon !conn !total !off = do
+  marr@(MutableByteArray marr#) <- PM.newPinnedByteArray total
+  interruptibleReceiveBoundedMutableByteArraySlice abandon conn total marr off >>= \case
+    Left err -> pure (Left err)
+    Right _ -> pure (Right (BI.PS (FP.ForeignPtr (byteArrayContents# (unsafeCoerce# marr#)) (FP.PlainPtr marr#)) 0 total))
+
+-- | Receive bytes using the given array. This can be interrupted
+--   by the completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArray ::
+     TVar Bool
+     -- ^ Interrupted. If this becomes 'True' give up and return
+     --   @'Left' 'ReceiveInterrupted'@.
+  -> Connection -- ^ Connection
+  -> MutableByteArray RealWorld -- ^ Buffer in which the data is going to be stored
+  -> IO (Either (ReceiveException 'Interruptible) Int) -- ^ Either a socket exception or the number of bytes read
+interruptibleReceiveMutableByteArray !abandon !conn !marr
+  = interruptibleReceiveMutableByteArraySlice abandon conn marr 0
+
+-- | Receive bytes using the given array and
+--   starting at the given offset. This can be interrupted by the
+--   completion of an 'STM' transaction.
+interruptibleReceiveMutableByteArraySlice ::
```

Comment:
This function is missing an argument that describes the length of the slice.

Reply:
isn't that the job of `interruptibleReceiveBoundedMutableByteArraySlice`?

Reply:
The `Bounded` variants receive up to a given number of bytes. The non-`Bounded` variants (which is what this function is) receive *exactly* a requested number of bytes by calling `recv` in a loop. Both variants take a length argument. There's a non-sliced variant that doesn't take a length argument, but that's not what is being defined here.

Reply:
isn't the length of the slice just the (length<sub>bytearray</sub> - offset)?

Reply:
oh, duh, slices don't always end at the end of the array. sorry.